File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];
        int product = 1;
        int n = nums.length;
        for(int i=0; i<n; i++) {
            if(i==0) {
                ans[i] = nums[i];
                continue;
            }
            ans[i] = ans[i-1] * nums[i];
        }
        for(int a: ans) {
            System.out.print(a + " ");
        }
        product = 1;
        for(int i=n-1; i>0; i--) {
            ans[i] = ans[i-1] * product;
            product *= nums[i];
        }
        ans[0] = product;
        return ans;
    }
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left = [1 for _ in range(n)]
        left[0] = nums[0]
        right = [1 for _ in range(n)]
        right[-1] = nums[-1]
        for i in range(1, n):
            left[i] = nums[i] * left[i-1]
        for i in range(n-2, -1, -1):
            right[i] = nums[i] * right[i+1]
        ans = [1 for _ in range(n)]
        ans[0] = right[1]
        ans[-1] = left[-2]
        for i in range(1, n-1):
            ans[i] = left[i-1] * right[i+1]
        return ans